// @sanity/code-input/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SUPPORTED_THEMES = exports.SUPPORTED_LANGUAGES = exports.PATH_LANGUAGE = exports.PATH_FILENAME = exports.PATH_CODE = exports.LANGUAGE_ALIASES = exports.DEFAULT_THEME = exports.ACE_SET_OPTIONS = exports.ACE_EDITOR_PROPS = void 0;
// NOTE: MAKE SURE THESE ALIGN WITH IMPORTS IN ./editorSupport
var SUPPORTED_LANGUAGES = [{
  title: 'Batch file',
  value: 'batchfile'
}, {
  title: 'C#',
  value: 'csharp'
}, {
  title: 'CSS',
  value: 'css'
}, {
  title: 'Go',
  value: 'golang'
}, {
  title: 'GROQ',
  value: 'groq'
}, {
  title: 'HTML',
  value: 'html'
}, {
  title: 'Java',
  value: 'java'
}, {
  title: 'JavaScript',
  value: 'javascript'
}, {
  title: 'JSON',
  value: 'json'
}, {
  title: 'JSX',
  value: 'jsx'
}, {
  title: 'Markdown',
  value: 'markdown'
}, {
  title: 'MySQL',
  value: 'mysql'
}, {
  title: 'PHP',
  value: 'php'
}, {
  title: 'Plain text',
  value: 'text'
}, {
  title: 'Python',
  value: 'python'
}, {
  title: 'Ruby',
  value: 'ruby'
}, {
  title: 'SASS',
  value: 'sass'
}, {
  title: 'SCSS',
  value: 'scss'
}, {
  title: 'sh',
  value: 'sh'
}, {
  title: 'TSX',
  value: 'tsx'
}, {
  title: 'TypeScript',
  value: 'typescript'
}, {
  title: 'XML',
  value: 'xml'
}, {
  title: 'YAML',
  value: 'yaml'
}];
exports.SUPPORTED_LANGUAGES = SUPPORTED_LANGUAGES;
var LANGUAGE_ALIASES = {
  js: 'javascript'
};
exports.LANGUAGE_ALIASES = LANGUAGE_ALIASES;
var SUPPORTED_THEMES = ['github', 'monokai', 'terminal', 'tomorrow'];
exports.SUPPORTED_THEMES = SUPPORTED_THEMES;
var DEFAULT_THEME = 'tomorrow';
exports.DEFAULT_THEME = DEFAULT_THEME;
var ACE_SET_OPTIONS = {
  useSoftTabs: true,
  navigateWithinSoftTabs: true
  /* note only supported by ace v1.2.7 or higher */

};
exports.ACE_SET_OPTIONS = ACE_SET_OPTIONS;
var ACE_EDITOR_PROPS = {
  $blockScrolling: true
};
exports.ACE_EDITOR_PROPS = ACE_EDITOR_PROPS;
var PATH_LANGUAGE = ['language'];
exports.PATH_LANGUAGE = PATH_LANGUAGE;
var PATH_CODE = ['code'];
exports.PATH_CODE = PATH_CODE;
var PATH_FILENAME = ['filename'];
exports.PATH_FILENAME = PATH_FILENAME;
