// @sanity/code-input/lib/createHighlightMarkers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createHighlightMarkers;
exports.highlightMarkersCSS = void 0;

var _styledComponents = require("styled-components");

var _templateObject;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var highlightMarkersCSS = (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  .ace_editor_markers_highlight {\n    position: absolute;\n    background-color: ", ";\n    opacity: 0.2;\n    width: 100% !important;\n    border-radius: 0 !important;\n  }\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.color.solid.primary.enabled.bg;
});
exports.highlightMarkersCSS = highlightMarkersCSS;

function createHighlightMarkers(rows) {
  return rows.map(row => ({
    startRow: Number(row) - 1,
    startCol: 0,
    endRow: Number(row) - 1,
    endCol: +Infinity,
    className: 'ace_editor_markers_highlight',
    type: 'screenLine',
    inFront: true
  }));
}
