// @sanity/code-input/lib/schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _CodeInput = _interopRequireDefault(require("./CodeInput"));

var _PreviewCode = _interopRequireDefault(require("./PreviewCode"));

var _getMedia = require("./getMedia");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Preview = props => {
  return /*#__PURE__*/_react.default.createElement(_PreviewCode.default, props);
};

var _default = {
  name: 'code',
  type: 'object',
  title: 'Code',
  inputComponent: _CodeInput.default,
  icon: _icons.CodeBlockIcon,
  fields: [{
    name: 'language',
    title: 'Language',
    type: 'string'
  }, {
    name: 'filename',
    title: 'Filename',
    type: 'string'
  }, {
    title: 'Code',
    name: 'code',
    type: 'text'
  }, {
    title: 'Highlighted lines',
    name: 'highlightedLines',
    type: 'array',
    of: [{
      type: 'number',
      title: 'Highlighted line'
    }]
  }],
  preview: {
    select: {
      language: 'language',
      code: 'code',
      filename: 'filename',
      highlightedLines: 'highlightedLines'
    },
    prepare: value => {
      return {
        title: value.filename || (value.language || 'unknown').toUpperCase(),
        media: (0, _getMedia.getMedia)(value === null || value === void 0 ? void 0 : value.language),
        extendedPreview: /*#__PURE__*/_react.default.createElement(Preview, {
          value: value
        })
      };
    }
  }
};
exports.default = _default;
